rcl/src/rcl/node.c

```diff
@@ -270,96 +277,83 @@ rcl_node_init(
     graph_guard_condition_options);
   if (ret != RCL_RET_OK) {
     // error message already set
-    goto fail;
+    goto free_graph_guard_condition;
   }
 
   // To capture all types from builtin topics and services, the type cache needs to be initialized
   // before any publishers/subscriptions/services/etc can be created
   ret = rcl_node_type_cache_init(node);
   if (ret != RCL_RET_OK) {
-    goto fail;
+    goto guard_condition_fini;
   }
 
-  // The initialization for the rosout publisher requires the node to be in initialized to a point
+  // The initialization for the rosout publisher requires the node to be initialized to a point
   // that it can create new topic publishers
   if (rcl_logging_rosout_enabled() && node->impl->options.enable_rosout) {
     ret = rcl_logging_rosout_init_publisher_for_node(node);
     if (ret != RCL_RET_OK) {
       // error message already set
-      goto fail;
+      goto node_type_cache_fini;
     }
   }
+
   RCUTILS_LOG_DEBUG_NAMED(ROS_PACKAGE_NAME, "Node initialized");
-  ret = RCL_RET_OK;
   TRACETOOLS_TRACEPOINT(
     rcl_node_init,
     (const void *)node,
     (const void *)rcl_node_get_rmw_handle(node),
     rcl_node_get_name(node),
     rcl_node_get_namespace(node));
-  goto cleanup;
-fail:
-  if (node->impl) {
-    if (rcl_logging_rosout_enabled() &&
-      node->impl->options.enable_rosout &&
-      node->impl->logger_name)
-    {
-      ret = rcl_logging_rosout_fini_publisher_for_node(node);
```

Comment:
Oh, yikes.  I missed that, good call.

Reply:
Oh wait, no.  Now that I look at this again, I removed this on purpose.  `rcl_logging_rosout_init_publisher_for_node` is the last substantial thing we do.  So there is no way to get into the error path after having called that.  Thus, we never need to clean it up.

Reply:
i think this assumption is based on implementation detail here? what if someone modifies `rcl_logging_rosout_init_publisher_for_node` later, which leads that to fail here possibly? according to API, i think we can keep `rcl_logging_rosout_fini_publisher_for_node` to be called?

btw,

https://github.com/ros2/rcl/blob/41b83c8ebfe8344a80c0093083b7abdfded79337/rcl/include/rcl/logging_rosout.h#L104

i cannot find `rcl_logging_destroy_rosout_publisher_for_node`, we probably need to change it into `rcl_logging_rosout_fini_publisher_for_node`.

Reply:
> i think this assumption is based on implementation detail here?

Yes and no.  All of the error handling in here is based on an implementation detail, namely that if a call fails it cleans up after itself.  If any of the init-style functions don't follow that, we'll probably leak or be in an inconsistent state anyway.  So even if `rcl_logging_rosout_init_publisher_for_node`, fails, it should clean up after itself.  There is just nothing (currently) that happens after it that would cause us to need to finalize it.

Reply:
understood, thanks!